extensions/sql/data/JSONSQLWriter.cpp

```diff
@@ -0,0 +1,156 @@
+  int countColumnsInMaxCollector = 0;
+
+  for (std::size_t i = 0; i != row.size(); ++i) {
+    const soci::column_properties & props = row.get_properties(i);
+
+    const auto& columnName = utils::toLower(props.get_name());
+
+    if (pMaxCollector_ && rowCount == 0 && pMaxCollector_->hasColumn(columnName)) {
+      countColumnsInMaxCollector++;
+    }
+
+    rapidjson::Value name;
+    name.SetString(props.get_name().c_str(), props.get_name().length(), alloc);
+
+    rapidjson::Value valueVal;
+
+    if (row.get_indicator(i) == soci::i_null) {
+      const std::string null = "NULL";
+      valueVal.SetString(null.c_str(), null.length(), alloc);
+    } else {
+      switch (const auto dataType = props.get_data_type()) {
+        case soci::data_type::dt_string: {
+          const auto value = std::string(row.get<std::string>(i));
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, '\'' + value + '\'');
+          }
+          valueVal.SetString(value.c_str(), value.length(), alloc);
+        }
+        break;
+        case soci::data_type::dt_double: {
+          const auto value = row.get<double>(i);
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, value);
+          }
+          valueVal.SetDouble(value);
+        }
+        break;
+        case soci::data_type::dt_integer: {
+          const auto value = row.get<int>(i);
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, value);
+          }
+          valueVal.SetInt(value);
+        }
+        break;
+        case soci::data_type::dt_long_long: {
+          const auto value = row.get<long long>(i);
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, value);
+          }
+          valueVal.SetInt64(value);
+        }
+        break;
+        case soci::data_type::dt_unsigned_long_long: {
+          const auto value = row.get<unsigned long long>(i);
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, value);
+          }
+          valueVal.SetUint64(value);
+        }
+        break;
+        case soci::data_type::dt_date: {
+          // It looks like design bug in soci - for `dt_date`, it returns std::tm, which doesn't have mmilliseconds, but DB 'datetime' has milliseconds.
+          // Don't know if it is possible to get a string representation for 'dt_date' type.
+          // The problem for maxCollector, if milliseconds value is not stored as a maximum, then when running query with 'select ... datetimeColumn > maxValue', 
+          // it will be always at least one record since DB has milliseconds "maxValue.milliseconds".
+          // For a workaround in the string representation for 'dt_date', add '999' for maxCollector (won't work for cases where time precision is important).
```

Comment:
(!) "won't work for cases where time precision is important" - I would argue that this is most cases using timestamps. If we even have the possibility to have two records inserted in the same second (assuming that date field is filled with `getdate()`, or something equivalent), we have the chance to get our read transaction between the two insertions and in this case we will lose the second record, which is not good.

I think getting a string representation for the `dt_date` would involve changing the query itself to cast the column to a string type or some ODBC transfer magic - it is not something SOCI seems to support.

If we can't get SOCI to do this, the only solution we have is to store all records within the last second (if there is a datetime Maximum-value column) and don't create FlowFiles from already processed records.